modules/ismp/clients/ismp-optimism/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+	}
+}
+
+impl<
+	H: IsmpHost + Send + Sync + Default + 'static,
+	T: pallet_ismp_host_executive::Config + pallet::Config + 'static,
+> Clone for OptimismConsensusClient<H, T>
+{
+	fn clone(&self) -> Self {
+		Self(core::marker::PhantomData)
+	}
+}
+
+impl<
+	H: IsmpHost + Send + Sync + Default + 'static,
+	T: pallet_ismp_host_executive::Config + pallet::Config + 'static,
+> ConsensusClient for OptimismConsensusClient<H, T>
+{
+	fn verify_consensus(
+		&self,
+		host: &dyn IsmpHost,
+		consensus_state_id: ConsensusStateId,
+		trusted_consensus_state: Vec<u8>,
+		consensus_proof: Vec<u8>,
+	) -> Result<(Vec<u8>, VerifiedCommitments), Error> {
+		let OptimismUpdate { state_machine_id, l1_height, proof } =
+			OptimismUpdate::decode(&mut &consensus_proof[..])
+				.map_err(|_| Error::Custom("Cannot decode optimism update".to_string()))?;
+
+		let mut consensus_state = ConsensusState::decode(&mut &trusted_consensus_state[..])
+			.map_err(|_| Error::Custom("Cannot decode trusted consensus state".to_string()))?;
+
+		let l1_state_machine_height =
+			StateMachineHeight { id: consensus_state.l1_state_machine_id, height: l1_height };
+
+		let l1_state_commitment = host.state_machine_commitment(l1_state_machine_height)?;
+		let state_root = l1_state_commitment.state_root;
+
+		let mut state_machine_map: BTreeMap<StateMachineId, Vec<StateCommitmentHeight>> =
+			BTreeMap::new();
+
+		match proof {
+			OptimismConsensusProof::OpL2Oracle(payload_proof) => {
+				if let Some(oracle_address) =
+					Pallet::<T>::state_machines_oracle_addresses(state_machine_id)
+				{
+					let state = verify_optimism_payload::<H>(
+						payload_proof,
+						state_root,
+						oracle_address,
+						consensus_state_id.clone(),
+					)?;
+
+					let state_commitment_height = StateCommitmentHeight {
+						commitment: state.commitment,
+						height: state.height.height,
+					};
+
+					let mut state_commitment_vec: Vec<StateCommitmentHeight> = Vec::new();
+					state_commitment_vec.push(state_commitment_height);
+					state_machine_map
+						.insert(consensus_state.l1_state_machine_id, state_commitment_vec);
+
+					consensus_state.finalized_height = state.height.height;
+				}
+			},
+			OptimismConsensusProof::OpFaultProofGames(dispute_proof) => {
+				if let Some((dispute_game_factory, respected_game_types)) =
+					Pallet::<T>::state_machines_dispute_game_factories_types(state_machine_id)
+				{
+					let state = verify_optimism_dispute_game_proof::<H>(
+						dispute_proof,
+						state_root,
+						dispute_game_factory,
+						respected_game_types,
+						consensus_state_id.clone(),
+					)?;
+
+					let state_commitment_height = StateCommitmentHeight {
+						commitment: state.commitment,
+						height: state.height.height,
+					};
+
+					let mut state_commitment_vec: Vec<StateCommitmentHeight> = Vec::new();
+					state_commitment_vec.push(state_commitment_height);
+					state_machine_map
+						.insert(consensus_state.l1_state_machine_id, state_commitment_vec);
```

Comment:
This should be the L2 state machine id
```suggestion
						.insert(consensus_state.l1_state_machine_id, state_commitment_vec);
```